src/DearImGui/Vulkan.hs

```diff
@@ -0,0 +1,162 @@
+  ( vulkanCtx )
+
+
+C.context ( Cpp.cppCtx <> C.funCtx <> vulkanCtx )
+C.include "imgui.h"
+C.include "backends/imgui_impl_vulkan.h"
+Cpp.using "namespace ImGui"
+
+
+data InitInfo =
+  InitInfo
+  { instance'      :: !Vulkan.Instance
+  , physicalDevice :: !Vulkan.PhysicalDevice
+  , device         :: !Vulkan.Device
+  , queueFamily    :: !Word32
+  , queue          :: !Vulkan.Queue
+  , pipelineCache  :: !Vulkan.PipelineCache
+  , descriptorPool :: !Vulkan.DescriptorPool
+  , subpass        :: !Word32
+  , minImageCount  :: !Word32
+  , imageCount     :: !Word32
+  , msaaSamples    :: !Vulkan.SampleCountFlagBits
+  , mbAllocator    :: Maybe Vulkan.AllocationCallbacks
+  , checkResult    :: Vulkan.Result -> IO ()
+  }
+
+-- | Wraps @ImGui_ImplVulkan_Init@ and @ImGui_ImplVulkan_Shutdown@.
+withVulkan :: MonadIO m => InitInfo -> Vulkan.RenderPass -> ( Bool -> m a ) -> m a
+withVulkan ( InitInfo {..} ) renderPass action = do
+  let
+    instancePtr :: Ptr Vulkan.Instance_T
+    instancePtr = Vulkan.instanceHandle instance'
+    physicalDevicePtr :: Ptr Vulkan.PhysicalDevice_T
+    physicalDevicePtr = Vulkan.physicalDeviceHandle physicalDevice
+    devicePtr :: Ptr Vulkan.Device_T
+    devicePtr = Vulkan.deviceHandle device
+    queuePtr :: Ptr Vulkan.Queue_T
+    queuePtr = Vulkan.queueHandle queue
+    withCallbacks :: ( Ptr Vulkan.AllocationCallbacks -> IO a ) -> IO a
+    withCallbacks f = case mbAllocator of
+      Nothing        -> f nullPtr
+      Just callbacks -> alloca ( \ ptr -> poke ptr callbacks *> f ptr )
+  checkResultFunPtr <- liftIO $ $( C.mkFunPtr [t| Vulkan.Result -> IO () |] ) checkResult
+  res <- liftIO $ withCallbacks \ callbacksPtr ->
+      [C.block| bool {
+        ImGui_ImplVulkan_InitInfo initInfo;
+        VkInstance instance = { $( VkInstance_T* instancePtr ) };
+        initInfo.Instance = instance;
+        VkPhysicalDevice physicalDevice = { $( VkPhysicalDevice_T* physicalDevicePtr ) };
+        initInfo.PhysicalDevice = physicalDevice;
+        VkDevice device = { $( VkDevice_T* devicePtr ) };
+        initInfo.Device = device;
+        initInfo.QueueFamily = $(uint32_t queueFamily);
+        VkQueue queue = { $( VkQueue_T* queuePtr ) };
+        initInfo.Queue = queue;
+        initInfo.PipelineCache = $(VkPipelineCache pipelineCache);
+        initInfo.DescriptorPool = $(VkDescriptorPool descriptorPool);
+        initInfo.Subpass = $(uint32_t subpass);
+        initInfo.MinImageCount = $(uint32_t minImageCount);
+        initInfo.ImageCount = $(uint32_t imageCount);
+        initInfo.MSAASamples = $(VkSampleCountFlagBits msaaSamples);
+        initInfo.Allocator = $(VkAllocationCallbacks* callbacksPtr);
+        initInfo.CheckVkResultFn = $( void (*checkResultFunPtr)(VkResult) );
+        return ImGui_ImplVulkan_Init(&initInfo, $(VkRenderPass renderPass) );
+      }|]
+  a <- action ( res /= 0 )
+  liftIO do
+    [C.exp| void { ImGui_ImplVulkan_Shutdown(); } |]
+    freeHaskellFunPtr checkResultFunPtr
```

Comment:
Should this be changed to be `bracket`? If `action` crashes, for example, we still want to call `freeHaskellFunPtr`

Reply:
OK... I'm not too familiar with the details, do we need an uninterruptibleMask, or is it OK to just use `Control.Exception.bracket`? And what if ImGui_ImplVulkan_ShutDown crashes?

Reply:
>  And what if ImGui_ImplVulkan_ShutDown crashes?

I think this is the crux of the question as to whether you should mask. I think just `bracket` would be fine here (though it's not a particular informed judgement). But it at least catches the case where an exception is thrown from `action`.

Reply:
I realised I can't use bracket as is because of the `MonadIO` action type. Should I use `UnliftIO`'s bracket?

Reply:
Ah, good point. Yea, let's go with `UnliftIO.bracket`.